[ty] Implement `from x.y import z` ensuring `x.y` resolves in the current file

AlexWaygood opened the pull request:
## Summary

Implement `from x.y import z` ensuring `x.y` resolves in the current file if a `y` attribute isn't otherwise available on `x`. This is implemented by adding these kinds of imports to `available_submodule_attributes` and lowering the priority of `available_submodule_attributes`.

Keeping the priority low ensures that things like `from .y import y` in an `__init__.py` reliable exposes the function/class `y` and not the submodule `y` (an extremely common idiom).

Fixes https://github.com/astral-sh/ty/issues/1654

## Test Plan

Snapshot tests.

AlexWaygood committed:
[ty] Add `from` imports to `imported_modules` *if* the module being imported is not relative to the current module

AlexWaygood committed:
tinker with `available_submodule_attributes` priority..

AlexWaygood committed:
.

AlexWaygood commented:
everything's great as long as you don't scroll down to the bit of the primer diff where the `prefect` diagnostics are listed

Gankra commented:
Taking a look at the code, the errors on prefect seem completely legitimate? The file doesn't import it and the `__init__.py`'s along the chain don't import it. These symbols being available must be genuine spooky-action-at-a-distance that we should probably discourage.

Gankra commented on crates/ty_python_semantic/src/semantic_index/builder.rs:
```diff
@@ -1513,33 +1517,46 @@ impl<'ast> Visitor<'ast> for SemanticIndexBuilder<'_, 'ast> {
-                            symbol.into(),
-                            ImportFromSubmoduleDefinitionNodeRef { node },
-                        );
+                    if self.current_scope().is_global() && node.module.is_some() {
```
Should we add the `is_global` check to `ast::Stmt::Import` too? (It not being there bothers me...)

AlexWaygood commented:
> Also did you ever try not-distinguishing-import-kind? i.e. even making `import ...` weak?

I _did_, but it directly contradicts some [longstanding tests](https://github.com/astral-sh/ruff/blob/main/crates/ty_python_semantic/resources/mdtest/import/tracking.md#attribute-overrides-submodule) that were added in https://github.com/astral-sh/ruff/pull/14946 sooo... I got scared :P

I _do_ think it makes sense that if a user has `import a.b` at the top of their module, they almost certainly expect the attribute `a.b` to resolve to the `b` submodule of the package `a`, whereas they won't necessarily expect the same thing for the `a.b` access if all they have in their module is `from a.b import c`... but it's also the case that in tie-breaking situations, we just don't know for sure which way the tie will _actually_ be broken, and I don't know if it would be clear to users why we're distinguishing between the two kinds of imports.

The longstanding tests were also added before we added any kind of ecosystem check, so I do think it would be interesting to see the ecosystem effect from not-distinguishing-import-kind as well. I'd like to try it separately, after I've debugged some of the stranger diagnostics on this branch a bit more.

AlexWaygood committed:
more

AlexWaygood committed:
more

AlexWaygood commented on crates/ty_python_semantic/src/types.rs:
```diff
@@ -11765,33 +11771,55 @@ impl<'db> ModuleLiteralType<'db> {
+                        ty
+                    }
+                })
+            })
```
it's extremely hard to write a regression test for this (though I might try harder later), but empirically this gets rid of a false-positive diagnostic on [this line](https://github.com/DataDog/dd-trace-py/blob/7e78fe5f172442f2379ea725fe4e5ea534f00395/ddtrace/contrib/internal/trace_utils.py#L56) that was ultimately due to us thinking the `importing_file` of `ddtrace.internal` was `ddtrace/__init__.py` rather than `ddtrace/contrib/internal/trace_utils.py`

AlexWaygood commented:
I can't see any "obviously wrong" diagnostics in the ecosystem anymore among the added or changed diagnostics. Nearly all the _new_ diagnostics are on prefect. As noted above, I'm not sure why these aren't emitted on `main`; they seem like true positives, as do all new diagnostics on other repos. This does get rid of diagnostics across the ecosystem overall now; it's just a question of whether this behaviour is explainable to users.

Still TODO is
- to just try out the bigger hammer of always giving `available_submodule_attributes` lowest priority, which would certainly be a simpler rule (and therefore easier to explain to users).
- https://github.com/astral-sh/ruff/pull/21583#discussion_r2554196826 should probably be fixed in a standalone PR
- try harder to write a standalone regression test for https://github.com/astral-sh/ruff/pull/21583#discussion_r2554271218
- figure out why none of those prefect diagnostics are reported on `main`

AlexWaygood committed:
allow `from` imports in nonglobal scopes to add available submodule attributes

AlexWaygood committed:
all the submodule attributes all the time

Gankra committed:
regression test

Gankra committed:
reapply fix

Gankra commented:
Interesting cases:

* [pwndbg explicitly shadows it's own submodule with `None` and then in an init subroutine unshadows it](https://github.com/pwndbg/pwndbg/blob/9763fc34370d0fab9ef35c36f3c6c560991dbec8/pwndbg/aglib/__init__.py#L18-L52) with explicit reference to it doing things "so mypy understands"
  * This is basically the entire diff in errors. This messes up pwndbg real bad. I guess this is the motivating case but boy I sure wish they just had an `if TYPE_CHECKING` or something instead...
*  [ Module `scipy.special` has no member `multigammaln` ](https://github.com/scipy/scipy/blob/1c36a69785d3d27ae279f3efbc4d66d7dd8acc7b/scipy/stats/_multivariate.py#L1821) even though [it's exported in `__all__` but maybe we get confused about the complex definition of it?](https://github.com/scipy/scipy/blob/1c36a69785d3d27ae279f3efbc4d66d7dd8acc7b/scipy/special/__init__.py#L829)
    * This appears to be another case of `__getattr__` bailing us out -- `scipy/__init__.py` has a `__getattr__` and previously we were falling into it. So. Same issue as `prefect`.
*  [Module `scipy.linalg.blas` has no member `dnrm2`](https://github.com/scipy/scipy/blob/1c36a69785d3d27ae279f3efbc4d66d7dd8acc7b/scipy/linalg/tests/test_blas.py#L1025)
    * ditto

Gankra commented:
scipy has the same magic lazy imports with `__getattr__` thing that prefect does, although it comes up a lot less.

*  [ Module `scipy.special` has no member `multigammaln` ](https://github.com/scipy/scipy/blob/1c36a69785d3d27ae279f3efbc4d66d7dd8acc7b/scipy/stats/_multivariate.py#L1821) even though [it's exported in `__all__` but maybe we get confused about the complex definition of it?](https://github.com/scipy/scipy/blob/1c36a69785d3d27ae279f3efbc4d66d7dd8acc7b/scipy/special/__init__.py#L829)
*  [Module `scipy.linalg.blas` has no member `dnrm2`](https://github.com/scipy/scipy/blob/1c36a69785d3d27ae279f3efbc4d66d7dd8acc7b/scipy/linalg/tests/test_blas.py#L1025)

Gankra commented:
```diff
- [warning] possibly-missing-attribute - Submodule `modin` may not be available as an attribute on module `pandera.typing`
+ [error] unresolved-attribute - Object of type `<module 'pandera.typing'> | <module 'pandera.typing'>` has no attribute `modin`
```

This is an unfortunate but kinda minor degradation (we could potentially add some diagnostic cleaning for it).

Gankra committed:
cleanup doc

Gankra committed:
fixup

Gankra reviewed (approved):
I guess I actually didn't end up changing the code itself, so, the code changes honestly look like what I'd expect.

AlexWaygood commented:
Thank you so much for picking this up, and in particular for experimenting with #21653 separately!

I'm surprised at how small the diff is on #21653 -- it's basically _all_ pwndbg, which is doing very strange and unusual things with its submodule imports (["submodule imports Georg" should be excluded as an outlier](https://www.huffingtonpost.co.uk/entry/spiders-georg-meme_n_5412861), etc.). So I wonder if we should just adopt the rule proposed by #21653, where submodule attributes always have lower priority than symbols defined in the `__init__.py` module namespace? The attraction of that is that it's a much simpler rule to explain to users; it's not great to have to explain a subtle distinction where we treat a submodule attribute introduced by `import a.b` as having a higher priority than a submdoule attribute introduced by `from a.b import c`.

As per https://github.com/astral-sh/ruff/pull/14946#discussion_r1884626442, that would mean that we'd be moving towards mypy's behaviour here and away from pyright's. But empirically, the ecosystem results appear to indicate that mypy's heuristic works better much more of the time than pyright's does.

I'm curious for what others think; don't make any changes here until other folks have had a chance to chime in 😅 The disadvantage of adopting the rule in #21653 is that you can get counterintuitive situations where things like this can happen:

```py
import a.b

reveal_type(a.b)  # revealed: Literal[1]
```

but honestly, at some point you just have to say that libraries need to stop shadowing their submodules with variables in `__init__.py` if they want type checkers to understand this stuff.
